// ent42/license.js
const { debugOut } = require('../../vxAssistCommon.js');
const { v4: uuidv4 } = require('uuid');
const fs = require('fs');

function defaultLicense() {
    return {
        licId: uuidv4(),
        revoked: false,
        shareable: true,
        consumer: null,
        expiration: Date.now() + 1000 * 60 * 60 * 24 * 365,

        modules: {
            openai: {
                models: ['', ''],
                limits: {
                    tokenUsed: 0,
                    tokenMax: 4096,
                }
            },
            huggingface: {
                models: ['', ''],
                limits: {
                    tokenUsed: 0,
                    tokenMax: 4096,
                }
            },
        },
    };
}

class Licensing {
    constructor() {
        this.licenses = [];
    }

    revoke(license) {
        license.revoked = true;

        this.licenses.forEach((l) => {
            if (l.licId === license.licId) {
                l.revoked = true;
            }
        });

        this.saveToStorage();
    }

    issue() {
        const license = defaultLicense();
        this.licenses.push(license);
        this.saveToStorage();

        return license;
    }

    claim(licId, consumer) {
        var license = this.getById(licId);
        var current = this.getByConsumer(consumer);

        if(!license) { 
            /** License key is invalid 
             */
            return false; 
        }

        if (license.consumer === consumer) {
            /** License key is already claimed by this user
             */
            return false;
        }

        if (!license.shareable && license.consumer !== null && license.consumer !== consumer) {
            /** License key is already claimed by a different user and is not shareable 
             */
            return false;
        }
        
        if(current){
            /** Remove & Remember previously claimed license
             */
            this.licenses = this.licenses.filter((l) => l.licId !== current.licId);
            license.previousLicense = current;
        }

        license.consumer = consumer;            
        this.saveToStorage();

        return true;
    }

    validate(license, consumer) {
        
    }

    getByConsumer(consumer) {
        return this.licenses.find((l) => l.consumer === consumer);

    }

    getById(licId) {
        return this.licenses.find((l) => l.licId === licId);
    }

    saveToStorage() {
        debugOut('saving licences to persistent storage');

        var dbo = {
            lastChanged: Date.now().toISOString(),
            licenses: this.licenses
        }

        fs.writeFileSync('licenseStore.json', JSON.stringify(dbo, null, 2), 'utf8');
    }

    loadFromStorage() {
        debugOut('loading licences from persistent storage');

        var dbo = JSON.parse(fs.readFileSync('licenseStore.json', 'utf8'));
        this.licenses = dbo.licenses;
    }
}

module.exports = {
    Licensing: new Licensing
};
